### PR TITLE
MobileNet - emre570

### PR DESCRIPTION
**Pull request by emre570**
Made MobileNet for inspection. This file includes:

- What is MobileNet and where we use
- Used convolution types
- Channel-wise Linear Bottleneck Layers
- Tensorflow and PyTorch implementations
- Downloadable model link

### CODE REVIEW

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+## Convolution Types on MobileNet
+
+By replacing regular convolutional layers with these depthwise separable convolutions and pointwise convolutions, MobileNet achieves high accuracy while minimizing computational overhead, making it well-suited for mobile devices and other resource-limited platforms. There are two types of convolutions used in MobileNet:
+
+### Depthwise Seperable Convolutions
+
+In traditional convolutional layers, each filter applies its weight across all input channels simultaneously. Depthwise separable convolutions break this down into two steps: depthwise convolution followed by pointwise convolution.  
+
+This step performs a convolution separately for each channel (a single color or feature) in the input image using a small filter (usually 3x3). The output of this step is the same size as the input but with fewer channels.
```
Comment (merveenoyan) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx:15]: How is this different computation/memory-wise instead of doing this in regular convolutions?

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+### Pointwise Seperable Convolutions
+
+This type of convolution applies a single filter (usually 1x1) across all channels in both input and output layers. It has fewer parameters than regular convolution and can be seen as an alternative to fully connected layers, making it suitable for mobile devices that have limited computational resources.
```
Comment (merveenoyan) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx:19]: How does 1x1 (kernel) conv work, you just take the pixels as is? isn't convolution calculated over a kernel? 

also regular convolution itself is better than fully connected layers itself so you could maybe compare to regular conv itself.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+### Channel-wise Linear Bottleneck Layers
+Channel-wise linear bottleneck layers help to further reduce the number of parameters and computational cost while maintaining high accuracy in image classification tasks.
+
+A channel-wise linear bottleneck layer consists of three main operations applied sequentially:  
+1. **Depthwise convolution:** This step performs a convolution separately for each channel (a single color or feature) in the input image using a small filter (usually 3x3). The output of this step is the same size as the input but with fewer channels.  
+2. **Batch normalization:** This operation normalizes the activation values across each channel, helping to stabilize the training process and improve generalization performance.  
+3. **Activation function:** Typically, a ReLU (Rectified Linear Unit) activation function is used after batch normalization to introduce non-linearity in the network.
```
Comment (merveenoyan) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx:31]: is ReLU added only for non-linearity or also to reduce vanishing gradients?

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+You can find implementation of MobileNet for both Tensorflow and PyTorch below:
+
+Tensorflow:
+```python
```
Comment (merveenoyan) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/mobilenet.mdx:36]: You can write framework dependent code like this so people can toggle in page directly:

<frameworkcontent>
<pt>
PyTorch content goes here
</pt>
<tf>
TensorFlow content goes here
</tf>

**Review by merveenoyan — COMMENTED** [review]
This is incredibly neatly written, very high quality content 🥹 ❤️  We can merge soon 😊

**Comment by emre570** [comment]
Thanks for your advices and I'm glad you liked it! I will edit it as soon as possible :)

**Comment by johko** [comment]
Hey @emre570, I didn't have the time to go through it in detail yet, but I realized you put the files in the wrong place. We got rid of the `course` folder and have all our content in the `chapters` folder now, sorry for the confusion

**Comment by emre570** [comment]
@johko I made some changes as you said, hope I made it right :)

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+#### Depthwise Separable Convolutions - Two-Step, Lighter Process:
+MobileNet does this process basic. It splits the big filter into two smaller, simpler steps:
+
+ - **Step 1 - Depthwise Convolution:** First, it uses a thin filter (like a single layer of a sponge) for each feature of the image
+   separately (like processing each color individually). This is less work because each filter is small and simple.
+   
+ - **Step 2 - Pointwise Convolution:** Then, it uses another small filter (just a tiny dot) to mix these features back together. This step is like taking a summary of what the first filters found.
+
+#### What this all means?
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:38]:
```suggestion
#### What does this all mean?
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+This step performs a convolution separately for each channel (a single color or feature) in the input image using a small filter (usually 3x3). The output of this step is the same size as the input but with fewer channels.
+
+### Pointwise Seperable Convolutions
+
+This type of convolution applies a single filter (usually 1x1) across all channels in both input and output layers. It has fewer parameters than regular convolution and can be seen as an alternative to fully connected layers, making it suitable for mobile devices that have limited computational resources.
+
+After depthwise convolution, this step combines the filtered outputs from previous steps using another 1x1 convolutional layer. This operation effectively aggregates the features learned by the depthwise convolutions into a smaller set of features, reducing the overall complexity while retaining important information.
+
+### Why We Use This Convolutions Instead of Regular Convolutions?
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:23]:
```suggestion
### Why Do We Use These Convolutions Instead of Regular Convolutions?
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+   separately (like processing each color individually). This is less work because each filter is small and simple.
+   
+ - **Step 2 - Pointwise Convolution:** Then, it uses another small filter (just a tiny dot) to mix these features back together. This step is like taking a summary of what the first filters found.
+
+#### What this all means?
+MobileNet uses these two smaller steps instead of one big step, it's like doing a lighter version of the work needed in regular convolutions. It's more efficient, especially on devices that aren't very powerful, like smartphones.
+
+With smaller filters, MobileNet doesn't need as much memory. It's like needing a smaller box to store all your tools, making it easier to fit into smaller devices.
+
+### How 1x1 Convolutions Work Compared to Normal Convolutions?
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:43]:
```suggestion
### How Do 1x1 Convolutions Work Compared to Normal Convolutions?
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  
+Certainly! Let's explain how a 1x1 convolution (kernel) works in comparison to normal convolutions in a straightforward manner:
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:45]:
```suggestion
Let's explain how a 1x1 convolution (kernel) works in comparison to normal convolutions in a straightforward manner:
```

Comment (MKhalusova) [reply chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:45]: "Certainly" implies yes/no question, and the title was a "how"

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+### Channel-wise Linear Bottleneck Layers
+Channel-wise linear bottleneck layers help to further reduce the number of parameters and computational cost while maintaining high accuracy in image classification tasks.
+
+A channel-wise linear bottleneck layer consists of three main operations applied sequentially:  
+1. **Depthwise convolution:** This step performs a convolution separately for each channel (a single color or feature) in the input image using a small filter (usually 3x3). The output of this step is the same size as the input but with fewer channels.  
+2. **Batch normalization:** This operation normalizes the activation values across each channel, helping to stabilize the training process and improve generalization performance.  
+3. **Activation function:** Typically, a ReLU (Rectified Linear Unit) activation function is used after batch normalization to introduce non-linearity in the network.
+
+### What ReLU does?
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:73]:
```suggestion
### What Does ReLU Do?
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        x = x.view(x.size(0), -1)
+        x = self.fc(x)
+
+        return x
+
+# Example of creating a MobileNet model
+mobilenet = MobileNet(num_classes=1000)
+print(mobilenet)
+```
+You can also find downloadable MobileNet model via this HuggingFace [link](https://huggingface.co/google/mobilenet_v2_1.0_224).
```
Comment (MKhalusova) [inline chapters/en/Unit 2 - Convolutional Neural Networks/mobilenetPT.mdx:174]:
```suggestion
You can also find a pretrained MobileNet model checkpoint on [HuggingFace Hub](https://huggingface.co/google/mobilenet_v2_1.0_224).
```

**Review by MKhalusova — COMMENTED** [review]
Thanks for a thorough overview of MobileNet. My main question is, why do we have two markdown files with the same content that only differ in the implementation example? 

When using doc-builder, we can combine content for different frameworks with a special tag, you can read more here - https://github.com/huggingface/doc-builder#framework-content.

```<frameworkcontent>
<pt>
PyTorch content goes here
</pt>
<tf>
TensorFlow content goes here
</tf>
<flax>
Flax content goes here
</flax>
</frameworkcontent>
```

This way, there's no need to duplicate the content.

**Comment by emre570** [comment]
Thanks for your suggestions. I will edit them ASAP as you said 😊

File: chapters/en/Unit 2 - Convolutional Neural Networks/vgg.mdx [hunk 11]
```diff
@@ -0,0 +1,92 @@
+
+# Very Deep Convolutional Networks for Large Scale Image Recognition (2014)
+
+
+## Introduction
+
+The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group -and hence named VGG- at Oxford University. The model demonstrated significant improvements over the past models at that time- to be specific 2014 Imagenet challange also known as ILSVRC.
```
Comment (johko) [inline chapters/en/Unit 2 - Convolutional Neural Networks/vgg.mdx:7]:
```suggestion
The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group -and hence named VGG- at Oxford University. The model demonstrated significant improvements over the past models at that time - specifically on the 2014 Imagenet challenge also known as ILSVRC.
```

**Review by johko — COMMENTED** [review]
I think Maria and me just reviewed at the same time, so maybe there is some overlap here :smile: 

But thank you for the thorough content, I just have the same remark as Maria about the two files with almost identical content.

But apart from that it looks really nice :+1:

**Comment by merveenoyan** [comment]
@emre570 this is very ready to be merged, can you solve the merge conflicts on README? should be trivial

**Comment by emre570** [comment]
I merged my branch and resolved conflicts, hope this works

**Review by merveenoyan — APPROVED** [review]

**Comment by merveenoyan** [comment]
Hey @emre570 if this is possible, I forgot to ask you to add how to use MobileNet through transformers or timm. Can you do that on a separate PR? Should be very short as it's inference

**Comment by emre570** [comment]
I will do my best
